caiman/source_extraction/cnmf/lazy_arrays.py

```diff
@@ -0,0 +1,439 @@
+
+        Returns:
+
+        """
+        pass
+
+    def __getitem__(
+            self,
+            item: Union[int, Tuple[slice_or_int_or_range]]
+    ):
+        if isinstance(item, int):
+            indexer = item
+
+        # numpy int scaler
+        elif isinstance(item, np.integer):
+            indexer = item.item()
+
+        # treat slice and range the same
+        elif isinstance(item, (slice, range)):
+            indexer = item
+
+        elif isinstance(item, tuple):
+            if len(item) > len(self.shape):
+                raise IndexError(
+                    f"Cannot index more dimensions than exist in the array. "
+                    f"You have tried to index with <{len(item)}> dimensions, "
+                    f"only <{len(self.shape)}> dimensions exist in the array"
+                )
+
+            indexer = item[0]
+
+        else:
+            raise IndexError(
+                f"You can index LazyArrays only using slice, int, or tuple of slice and int, "
+                f"you have passed a: <{type(item)}>"
+            )
+
+        # treat slice and range the same
+        if isinstance(indexer, (slice, range)):
+            start = indexer.start
+            stop = indexer.stop
+            step = indexer.step
+
+            if start is not None:
+                if start > self.n_frames:
+                    raise IndexError(f"Cannot index beyond `n_frames`.\n"
+                                     f"Desired frame start index of <{start}> "
+                                     f"lies beyond `n_frames` <{self.n_frames}>")
+            if stop is not None:
+                if stop > self.n_frames:
+                    raise IndexError(f"Cannot index beyond `n_frames`.\n"
+                                     f"Desired frame stop index of <{stop}> "
+                                     f"lies beyond `n_frames` <{self.n_frames}>")
+
+            if step is None:
+                step = 1
+
+            # convert indexer to slice if it was a range, allows things like decord.VideoReader slicing
+            indexer = slice(start, stop, step)  # in case it was a range object
+
+            # dimension_0 is always time
+            frames = self._compute_at_indices(indexer)
+
+            # index the remaining dims after lazy computing the frame(s)
+            if isinstance(item, tuple):
+                if len(item) == 2:
+                    return frames[:, item[1]]
+                elif len(item) == 3:
+                    return frames[:, item[1], item[2]]
+
+            else:
+                return frames
+
+        elif isinstance(indexer, int):
+            return self._compute_at_indices(indexer)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__} @{hex(id(self))}\n" \
+               f"{self.__class__.__doc__}\n" \
+               f"Frames are computed only upon indexing\n" \
+               f"shape [frames, x, y]: {self.shape}\n"
+
+
+class LazyArrayRCM(LazyArray):
```

Comment:
I wonder if spelling it out is preferable. It's already kinda long though; LazyArrayReconstructedMovie might be a bit ridiculous.

Reply:
On the naming options: as I mentioned in my general comment, I am not a fan of the abbreviations RCM etc. I think we should probably cut the mention of array in the names, and that they are lazy. E.g., we could just do `ReconstructedMovie`, `ReconstructedBackground`, `ReconstructedResiduals`. Refer to the objects by what they are, not their data type or back end details -- we don't name zarr arrays in terms of their ability to be randomly accessed.  Yes, these details are obviously really important for our goals, but once these things have inherited from `LazyArray()`, do we want our naming conventions to be captive to this?

Reply:
Let's try to find a name that's clear but not so long. RCM may be too short, but ReconstructedMovie feels long enough to be Java-ish.

Reply:
These are well within the bounds of what I'd be comfortable with lengthwise, but if anyone can think of apt shorter names I'd be fine if they don't have hard to understand abbreviations. Also, classes are typically not instantiated that much, and people can name the instances shorter little nubs as they like.

Reply:
I'm not happy with those lengths. If you ever look at code in Java it has those kinds of long names and there is a lot of typing.

Maybe ReconstMov, ReconstBkg, ReconstResids?

Maybe even RcMov, RcBkg, RcResids?